pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -59,43 +59,58 @@ cdef inline object create_timestamp_from_ts(int64_t value,
 
 
 def round_ns(values, rounder, freq):
+
     """
     Applies rounding function at given frequency
 
     Parameters
     ----------
-    values : int, :obj:`ndarray`
-    rounder : function
+    values : :obj:`ndarray`
+    rounder : function, eg. 'Ceil', 'Floor', 'round'
     freq : str, obj
 
     Returns
     -------
-    int or :obj:`ndarray`
+    :obj:`ndarray`
     """
+    def _round_non_int_multiple(value, unit):
+
+        if unit < 1000:
+            # for nano rounding, work with the last 6 digits separately
+            # due to float precision
+            buff = 1000000
+            r = (buff * (value // buff) + unit *
+                 (rounder((value % buff) * (1 / float(unit)))).astype('i8'))
+        else:
+            if unit % 1000 != 0:
+                msg = 'Precision will be lost using frequency: {}'
+                warnings.warn(msg.format(freq))
+
+            # GH19206
+            # to deal with round-off when unit is large
+            if unit >= 1e9:
+                divisor = 10 ** int(np.log10(unit / 1e7))
+            else:
+                divisor = 10
+
+            r = (unit * rounder((value * (divisor / float(unit))) / divisor)
+                 .astype('i8'))
+
+        return r
+
     from pandas.tseries.frequencies import to_offset
     unit = to_offset(freq).nanos
-    if unit < 1000:
-        # for nano rounding, work with the last 6 digits separately
-        # due to float precision
-        buff = 1000000
-        r = (buff * (values // buff) + unit *
-             (rounder((values % buff) * (1 / float(unit)))).astype('i8'))
-    else:
-        if unit % 1000 != 0:
-            msg = 'Precision will be lost using frequency: {}'
-            warnings.warn(msg.format(freq))
-
-        # GH19206
-        # to deal with round-off when unit is large
-        if unit >= 1e9:
-            divisor = 10 ** int(np.log10(unit / 1e7))
-        else:
-            divisor = 10
 
-        r = (unit * rounder((values * (divisor / float(unit))) / divisor)
-             .astype('i8'))
+    values = np.copy(values)
```

Comment:
Why do we need a copy here?

Reply:
This is to handle the case where 'NaT' exists in the DateTimeIndex. Removing it will cause `test_ceil_floor_edge` in `test_scalar_compact.py` to fail.

I found that datetimelike.py `self.hasnans` in `_maybe_mask_results`will return False if we don't do the copy, by copying we ensure that we arn't referencing the base of the input array. Think that is the issue here, what do you think?

Reply:
cc @jreback.

Thanks for the investigation. That sounds reasonable; but I am not too familiar with `nan`/`Nat` ops with respect to references.

Reply:
you are mutating in place, so you DO need to copy.

that's fine. though pls use ``values.copy()``